src/grid.rs

```diff
@@ -0,0 +1,207 @@
+extern crate rand as rand_crate;  // Alias the rand crate
+use rand_crate::prelude::*;  // Use the alias here
+
+use crate::cell::Cell;
+
+#[derive(Debug)]
+pub struct Grid{
+    pub cells: Vec<Vec<Cell>>
+}
+
+impl Grid{
+    fn rows(&self) -> usize {
+        self.cells.len()
+    }
+
+    fn cols(&self) -> usize {
+        self.cells[0].len()
+    }
+
+    // New grid with dead cells
+    pub fn new(rows: usize, cols: usize) -> Self {
+        Grid { 
+            cells: vec![vec![Cell::Dead; cols]; rows],
+        }
+    }
+
+    pub fn new_from_cells(cells: Vec<Vec<Cell>>) -> Self {
+        Grid { 
+            cells
+        }
+    }
+
+    pub fn new_random(rows: usize, cols: usize) -> Self {
+        let mut rng = rand_crate::thread_rng();
+        let mut cells = vec![vec![Cell::Dead; cols]; rows];
+        
+        for i in 0..rows {
+            for j in 0..cols {
+                if rng.gen_bool(0.5) {
+                    cells[i][j] = Cell::Alive;
+                }
+            }
+        }
+
+        Grid { cells }
+    }
+
+    // Gets neighbors of a given position and counts how many of them are alive cells.
+    pub fn count_alive_neighbors(&self, x: usize, y: usize) -> u8 {
+        let mut count = 0;
+
+        let neighbors = self.get_neighbors(x, y);
+
+        for neighbor in neighbors{
+            if self.cells[neighbor.0][neighbor.1] == Cell::Alive {
+                count += 1;
+            }
+        }
+        
+        count
+    }
+
+    // Get all neighbors of a given position (not the actual Cell velues, but their positions in the grid)
+    // Handles the case in which there might be a position with fewer neighbors. Like [0,0]
+    pub fn get_neighbors(&self, x:usize, y:usize) -> Vec<(usize, usize)>{
+        let mut neighbors: Vec<(usize, usize)> = Vec::new();
+
+        // Relative positions
+        let deltas = [
+            (-1, -1), (-1, 0), (-1, 1), 
+            (0, -1), /*(0,0),*/ (0, 1), 
+            (1, -1), (1, 0), (1, 1),
+        ];
+
+        for (dx, dy) in &deltas {
+            // I use isize because result can be negative.
+            let new_x = x as isize + dx;
+            let new_y = y as isize + dy;
+
+            // Only push neighbor to neighbors if it's position is valid
+            if new_x >= 0 && new_x < self.rows() as isize && 
```

Comment:
I would move this check `new_x >= 0 && new_x < self.rows() as isize && new_y >= 0 && new_y < self.cols() as isize` to be an abstraction, for example something called `position_is_out_of_bounds` .

Reply:
Done! I implemented it as a private method in Grid. Do you think that's ok?

Reply:
Yeah thats great!